### DISCUSSION TITLE
Http extension not connecting with my local development project url.

### DISCUSSION THREAD

**Discussion by HenriqueRamos13** [supabase/supabase#13789] [Questions]
Hi, on this supabase tutorial [https://supabase.com/docs/guides/getting-started/tutorials/with-nextjs#storage-management](https://supabase.com/docs/guides/getting-started/tutorials/with-nextjs#storage-management) Im trying to delete the user profile images when a new one is uploaded, but im always getting this error on local development (i didnt tried on production): `{statusCode: '500', error: 'Database Error', message: 'Failed to connect to localhost port 54321: Connection refused'}`

I have this part of the code:

```
create extension http
with
  schema extensions;

create or replace function delete_storage_object(bucket text, object text, out status int, out content text)
returns record
language 'plpgsql'
security definer
as $$
declare
  project_url text := 'http://localhost:54321';
  service_role_key text := 'my_key_hide'; --  full access needed
  url text := project_url||'/storage/v1/object/'||bucket||'/'||object;
begin
  select
      into status, content
           result.status::int, result.content::text
      FROM extensions.http((
            'DELETE',
            url,
            ARRAY[extensions.http_header('authorization','Bearer '||service_role_key)],
            NULL,
            NULL
      )::extensions.http_request) as result;
end;
$$;
```

Shoul I put another project_url? Or make any configuration to let my database connect on my project? 

Edit: I tested it on production and worked. But on localhost still not working.

**Comment by GaryAustin1** [comment]
I don't do local dev, but the storage-api github shows this as an example URL
`curl --location --request POST 'http://localhost:5000/bucket' \` where they are accessing a bucket.
Based on that I would say you don't use `/storage/v1` on local hosted.  Instead `......5xxxx/object/bucketname/pathname`.

**Reply by HenriqueRamos13** [reply]
Still receiving `{statusCode: '500', error: 'Database Error', message: 'Failed to connect to localhost port 5000: Connection refused'}`, i tried the 5000 port and the 5432 that shown here:

```
Started supabase local development setup.

         API URL: http://localhost:54321
          DB URL: postgresql://postgres:postgres@localhost:54322/postgres
      Studio URL: http://localhost:54323
    Inbucket URL: http://localhost:54324
      JWT secret: super-secret-jwt-token-with-at-least-32-characters-long
        anon key: eyJhb
        ...
```

I modified my code to:

```
...
declare
  project_url text := 'http://localhost:54321';
  service_role_key text := 'eyJhb...'; --  full access needed
  url text := project_url||'/object/'||bucket||'/'||object;
  http_enabled boolean := true;
begin
...
```

Other thing is: when i upload a file to the storage, the url that supabase uses to make the request (looking on the browser developer tools) is:

`Request URL: http://localhost:54321/storage/v1/object/profiles/dddda9b5-3109-42d7-a729-317aef45cb47/1681772103364.jpeg`

So i think i need still using the /storage/v1 even on development local

**Reply by GaryAustin1** [reply]
Yeah I would agree then based on the url being generated by your supabase.  The docs in the repository seem a bit dated anyway.

![image](https://user-images.githubusercontent.com/54564956/232628787-1d467681-53d2-4f3d-b237-fe6bf4372137.png)

**Reply by GaryAustin1** [reply]
You should be able to use that exact URL you got from the dev console in your delete code as a test.

**Reply by HenriqueRamos13** [reply]
I look into the html to see what url the html is using and i got that:

`<img src="http://localhost:54321/storage/v1/object/public/profiles/dddda9b5-3109-42d7-a729-317aef45cb47/1681772080586.jpeg" alt="profile">`

When i do a post request to upload a new file the network tab show: 

http://localhost:54321/storage/v1/object/profiles/dddda9b5-3109-42d7-a729-317aef45cb47/1681772080586.jpeg

I tried to change the code on database to add the 'public' directory:

`url text := project_url||'/storage/v1/object/public'||bucket||'/'||object;` 

i logged the url that the db get and it shows: `"url: http://localhost:54321/storage/v1/object/profiles/dddda9b5-3109-42d7-a729-317aef45cb47/1681772080586.jpeg"`

but still getting: `{"statusCode":"500","error":"Database Error","message":"Failed to connect to localhost port 54321: Connection refused"}`

What is the chance to be a cors error, service-role error or something like that?

**Reply by HenriqueRamos13** [reply]
i tried on my terminal and get success:

`curl --location --request DELETE 'http://localhost:54321/storage/v1/object/profiles/dddda9b5-3109-42d7-a729-317aef45cb47/1681772080586.jpeg' --header 'authorization: Bearer secret_token'`

`
{"message":"Successfully deleted"}
`

the problem looks on the database now, some missconfiguration?

**Reply by GaryAustin1** [reply]
I guess there could be something on routing from the database to the storage.  Sorry got nothing else for you.

**Reply by HenriqueRamos13** [reply]
@GaryAustin1 i tried to swap to pg_net like in your repository code, but i always get a timeout,.

**Comment by HenriqueRamos13** [comment]
Btw, i tested it on production and worked. But on localhost still not working.

**Comment by fourcube** [comment]
@HenriqueRamos13 since postgres and the API gateway (kong) are running in separate containers, connecting via the project URL that `supabase status` gives you won't work on a local setup. The postgres container cannot connect to the API gateway via localhost:54321. 

## This worked for me

Find the name of your supabase kong container:

`docker ps | grep supabase_kong`

And put it as the hostname in the project_url variable. You also have to use port 8000 because 54321 is the mapped port on your local machine.

`project_url text := 'http://supabase_kong_<your_project_name>:8000';`

**Reply by 0xemc** [reply]
This was very helpful. Since the intention for local development is to match the functionality of the remote, these kinds of differences should definitely be documented more clearly.

**Comment by jziggas** [comment]
i just ran into this same issue. I can make a successful request using the same URL and key via Postman, but Postgres complains of a refused connection. I began wondering if this was either a bug or something to do with the Docker network, and found this thread.

And it turns out the solution suggested by @fourcube is correct! Changing the URL from http://localhost:54321 to http://supabase_kong_<your_project_name>:8000 I now have working image uploads. Thanks, fourcube.

This kind of thing really needs to be added to the documentation for local development on the Supabase site. There is currently none. And there are too many gotchas currently when trying to work from your machine, at feature parity, before deploying code to the hosted platform. This is just one of several I have run into the past few weeks.

**Reply by eRuaro** [reply]
Alternatively, `http://host.docker.internal:54321` also works! Same thing also applies if you're connecting to the DB via an ORM. The connection string should look something like this: `postgresql://postgres:postgres@host.docker.internal:54322/postgres` or `postgres://postgres:postgres@host.docker.internal:54322/postgres` depending on the ORM you're using.
